_rsrch/cidiface/README.md

```diff
@@ -129,6 +136,23 @@ if a situation is at the scale where it's become important to mind whether
 or not pointers are a performance impact, then that situation also
 is one where you have to think twice before using interfaces.
 
+### struct wrappers can be used in place of typedefs with zero overhead
+
+See `TestSizeOf`.
+
+Using the `unsafe.Sizeof` feature to inspect what the Go runtime thinks,
+we can see that `type Foo string` and `type Foo struct{x string}` consume
+precisely the same amount of memory.
+
+This is interesting because it means we can choose between either
+type definition with no significant overhead anywhere we use it:
+thus, we can choose freely between Option B and Option F based on which
+we feel is more pleasant to work with.
+
+Option F (a struct wrapper) means we can prevent casting into our Cid type.
+Option B (typedef string) can be declared a `const`.
+Are there any other concerns that would separate the two choices?
```

Comment:
The struct wrapper allows us to add additional metadata/offsets later if we want.

Reply:
People would still compare against "" and cast to a string/bytes.

Reply:
The former should be rejected in code review based on the existence of `cid.Zero` and introduced universally when we do the refactor; the latter would be a *bug*.

If we're seriously worried about these, then we should 100% hands-down prefer Option F over B, because more than precisely zero instances of casting to Cid from outside the cid package would strike me as acceptable code quality.

Reply:
> The former should be rejected in code review based on the existence of cid.Zero and introduced universally when we do the refactor; the latter would be a bug.

That's not an assumption we can make, generally. Users (not in our projects) *will* cast. It's really the *primary* reason to use a type alias over a struct.

Reply:
After thinking about this, the face that a struct value can not be declared a const (another misfeature of go in my option) might have a performance impact when comparing to the zero value (`cid.Nil`).  Since the zero value is not a const the go compiler can't optimize on that fact and may have to do a load from memory.

Reply:
> But I doubt it's a serious performance thing. 

Probably not.  But if we commit to making it a struct we could just use cid.Cid{}.  It is just a clear and avoids the problem altogether.

Reply:
I don't think it'll make a difference, as long as we never modify it.

Reply:
For the choice of the constant I am going with `Nil`.  My second choice would be `Empty`, and `Zero` the third.  `Nil` to me is neutral and also implies an invalid value, of which an empty Cid{} struct is (it can't be used for anything).  `Empty` to me implies a valid but empty container, but I am okay with it if `Nil` is causing that much of a problem.  `Zero` to me implies a numeric value, of which a Cid is not.  I understand the notion of a zero-value in compiler terminology, but `Zero` still strikes me as an odd choice, `ZeroValue` may be a _bit_ better, but if `Nil` is no good I still rather go with `Empty`.

I didn't realize that using struct initializers in the middle of equality was so verbose.  However, it was never my intent to use them in equality.  Instead I provided an `IsNil` method as c.IsNil() is nicer than c == cid.Nil or c == (cid.Cid{}).  However elected to still provide `var Nil = Cid{}` so it can be used if desired.

Reply:
If we're going to make speculations about whether comparison to a var is going to be a performance issue, then shouldn't proposing an equality *method* have at least as much burden to prove itself a non-issue before we consider that a confirmed design choice?

I disagree completely about naming a variable of non-pointer type "Nil".  That is extremely unidiomatic Go because it conflicts with the language's own definitions of the term, and in discussion elsewhere I've already linked the part of the language spec which defines zero versus nil.  But this PR is probably not the topical place to relitigate this naming, either, seeing as how the term is not used anywhere in the diff body we're discussing.

Reply:
I'm strongly in favor of `Nil` as a constant. While `Zero` is technically correct, it's confusing.

However, as a method, I'd be fine with `IsEmpty` (or `IsZero` but I'd rather not).

As for the method, golang is pretty good about inlining one-line methods so it should be fine. After doing a bit of decompiling, it looks like we should go with:

```func
func (c Cid) IsNil() bool {
    return c.string == ""
}
```

`c == (Cid{})` seems to cost us an extra xor zeroing something out. `c == Nil` costs us a mov.

See: https://go.godbolt.org/z/CHyhZl

Reply:
@Stebalien the `IsNilString` one looks wrong. The compiler is getting really smart there and looks like its comparing the empty string to itself, and calling that the output.

Reply:
(discussed on slack)

The assembly was correct. The generated instruction was `testq AX, AX` which tests `AX & AX` to see if it's non-nil.